### PR TITLE
Snarky/cs chaining safety check

### LINKED ISSUE #13812: Enhance snarky to detect witness inconsistencies

### PR DESCRIPTION
**Pull request by jspada**
Closes #13812

### CODE REVIEW

File: src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml [hunk 1]
```diff
@@ -986,8 +1014,22 @@ end = struct
...
         sys.next_row <- sys.next_row + 1 ;
+
+        (* Check consistency of candidate row against row requirements *)
+        Array.iteri sys.required_witness_row ~f:(fun col value ->
+            match value with
+            | None ->
+                ()
+            | Some target ->
+                if Stdlib.(vars.(col) <> value) then (
+                  printf "Invalid witness value in row %d column %d\n" (List.length sys.rows_rev) col ;
```
[the remark this answers is not in the document]
Comment (jspada) [reply src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml]: Thanks

File: src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml [hunk 2]
```diff
@@ -986,8 +1014,23 @@ end = struct
...
         sys.next_row <- sys.next_row + 1 ;
+
+        (* Check consistency of candidate row against row requirements *)
+        Array.iteri sys.required_witness_row ~f:(fun col value ->
+            match value with
+            | None ->
+                ()
+            | Some target ->
+                if Stdlib.(vars.(col) <> value) then (
+                  printf "Invalid witness value in row %d column %d\n"
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml]: Likewise here, `failwithf`

File: src/lib/crypto/kimchi_backend/gadgets/bitwise.ml [hunk 3]
```diff
@@ -705,6 +708,16 @@ let%test_unit "bitwise xor gadget" =
       let cs, _proof_keypair, _proof =
         Runner.generate_and_verify_proof ?cs (fun () ->
             let open Runner.Impl in
+            (* Create half a generic to force a possible generic in the middle *)
```
Comment (mrmr1993) [inline src/lib/crypto/kimchi_backend/gadgets/bitwise.ml:711]: Unrelated, revert.

**Review by mrmr1993 — APPROVED** [review]

**Review by querolita — APPROVED** [review]
Accepting, but note that this work is not complete without support for Rotation which is in https://github.com/MinaProtocol/mina/pull/13839 and also https://github.com/MinaProtocol/mina/pull/13837

Those shall be merged to develop after the current PR though.

**Pull request closed without merging**